File: rdkit-sys/src/bridge/ro_mol.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    #[repr(i32)]
    #[derive(Debug, PartialEq)]
    pub enum HybridizationType {
        UNSPECIFIED,
        S,
        SP,
        SP2,
        SP3,
        SP2D,
        SP3D,
        SP3D2,
        OTHER,
    }

    unsafe extern "C++" {
        include!("wrapper/include/ro_mol.h");

        pub type ROMol;
        pub type ExplicitBitVect = crate::fingerprint_ffi::ExplicitBitVect;
        pub type SmilesParserParams;
        pub type Atom;
        pub type HybridizationType;

        pub type MolSanitizeException;
        pub type MolSanitizeExceptionUniquePtr; //  = UniquePtr<MolSanitizeException>;

        pub fn copy_mol(mol: &SharedPtr<ROMol>) -> SharedPtr<ROMol>;

        pub fn smiles_to_mol(smi: &CxxString) -> Result<SharedPtr<ROMol>>;

        pub fn smiles_to_mol_with_params(
            smi: &CxxString,
            params: &SharedPtr<SmilesParserParams>,
        ) -> Result<SharedPtr<ROMol>>;
        pub fn new_smiles_parser_params() -> SharedPtr<SmilesParserParams>;
        pub fn smiles_parser_params_set_sanitize(
            ptr: &SharedPtr<SmilesParserParams>,
            sanitize: bool,
        );
        pub fn smiles_parser_params_get_sanitize(ptr: &SharedPtr<SmilesParserParams>) -> bool;

        pub fn mol_to_smiles(mol: &SharedPtr<ROMol>) -> String;

        pub fn mol_to_molblock(mol: &SharedPtr<ROMol>) -> String;

        pub fn detect_chemistry_problems(
            mol: &SharedPtr<ROMol>,
        ) -> UniquePtr<CxxVector<MolSanitizeExceptionUniquePtr>>;

        pub fn mol_sanitize_exception_type(
            mol_sanitize_exception: &MolSanitizeExceptionUniquePtr,
        ) -> String;

        pub fn atom_sanitize_exception_get_atom_idx(
            mol_sanitize_exception: &MolSanitizeExceptionUniquePtr,
        ) -> u32;

        pub fn get_num_atoms(mol: &SharedPtr<ROMol>, onlyExplicit: bool) -> u32;
        pub fn get_atom_with_idx(mol: &mut SharedPtr<ROMol>, idx: u32) -> Pin<&mut Atom>;
        pub fn get_symbol(atom: Pin<&Atom>) -> String;
        pub fn get_is_aromatic(atom: Pin<&Atom>) -> bool;
        pub fn get_atomic_num(atom: Pin<&Atom>) -> i32;
        pub fn get_formal_charge(atom: Pin<&Atom>) -> i32;
        pub fn get_total_num_hs(atom: Pin<&Atom>) -> u32;
        pub fn get_total_valence(atom: Pin<&Atom>) -> u32;
        pub fn set_formal_charge(atom: Pin<&mut Atom>, what: i32);
        pub fn set_num_explicit_hs(atom: Pin<&mut Atom>, what: i32);
        pub fn atom_update_property_cache(atom: Pin<&mut Atom>, strict: bool) -> Result<()>;
        pub fn atom_set_hybridization(atom: Pin<&mut Atom>, what: HybridizationType);
        pub fn atom_get_hybridization(atom: Pin<&Atom>) -> HybridizationType;

        pub fn ro_mol_update_property_cache(atom: &mut SharedPtr<ROMol>, strict: bool);

        pub fn set_int_prop(atom: Pin<&mut Atom>, key: &CxxString, value: i32);
        pub fn get_int_prop(atom: Pin<&Atom>, key: &CxxString) -> Result<i32>;
        pub fn set_float_prop(atom: Pin<&mut Atom>, key: &CxxString, value: f64);
        pub fn get_float_prop(atom: Pin<&Atom>, key: &CxxString) -> Result<f64>;
        pub fn set_bool_prop(atom: Pin<&mut Atom>, key: &CxxString, value: bool);
        pub fn get_bool_prop(atom: Pin<&Atom>, key: &CxxString) -> Result<bool>;
        pub fn set_prop(atom: Pin<&mut Atom>, key: &CxxString, value: &CxxString);
        pub fn get_prop(atom: Pin<&Atom>, key: &CxxString) -> Result<String>;
        pub fn get_num_radical_electrons(atom: Pin<&Atom>) -> u32;

        //Returns the degree of the atom in the molecule.
        pub fn get_degree(atom: Pin<&Atom>) -> u32;
    }
}

File: rdkit-sys/wrapper/src/ro_mol.cc
#include "rust/cxx.h"
#include <DataStructs/ExplicitBitVect.h>
#include <GraphMol/FileParsers/FileWriters.h>
#include <GraphMol/Fingerprints/Fingerprints.h>
#include <GraphMol/GraphMol.h>
#include <GraphMol/MolOps.h>
#include <GraphMol/MolStandardize/Tautomer.h>
#include <GraphMol/SmilesParse/SmilesParse.h>
#include <GraphMol/SmilesParse/SmilesWrite.h>

#include <iostream>

namespace RDKit {
using ExplicitBitVect = ::ExplicitBitVect;

std::shared_ptr<ROMol> copy_mol(const std::shared_ptr<ROMol> &mol) { return std::shared_ptr<ROMol>(new ROMol(*mol)); }

std::shared_ptr<ROMol> smiles_to_mol(const std::string &smiles) {
	ROMol *mol = SmilesToMol(smiles);

	return std::shared_ptr<ROMol>(mol);
}

rust::String mol_to_smiles(const std::shared_ptr<ROMol> &mol) { return MolToSmiles(*mol); }

rust::String mol_to_molblock(const std::shared_ptr<ROMol> &mol) { return MolToMolBlock(*mol); }

std::shared_ptr<ROMol> smiles_to_mol_with_params(const std::string &smiles,
                                                 const std::shared_ptr<SmilesParserParams> &params) {
	ROMol *mol = SmilesToMol(smiles, *params);

	return std::shared_ptr<ROMol>(mol);
}
std::shared_ptr<SmilesParserParams> new_smiles_parser_params() {
	return std::shared_ptr<SmilesParserParams>(new SmilesParserParams());
}
void smiles_parser_params_set_sanitize(const std::shared_ptr<SmilesParserParams> &params, bool sanitize) {
	params->sanitize = sanitize;
}
bool smiles_parser_params_get_sanitize(const std::shared_ptr<SmilesParserParams> &params) { return params->sanitize; }

using MolSanitizeExceptionUniquePtr = std::unique_ptr<MolSanitizeException>;
std::unique_ptr<std::vector<MolSanitizeExceptionUniquePtr>>
detect_chemistry_problems(const std::shared_ptr<ROMol> &mol) {
	std::vector<MolSanitizeExceptionUniquePtr> exceptions         = RDKit::MolOps::detectChemistryProblems(*mol);
	std::vector<MolSanitizeExceptionUniquePtr> *heaped_exceptions = new std::vector<MolSanitizeExceptionUniquePtr>();

	auto s = exceptions.size();
	heaped_exceptions->reserve(s);
	for (int i = 0; i < s; i++) { heaped_exceptions->push_back(std::move(exceptions[i])); }

	return std::unique_ptr<std::vector<MolSanitizeExceptionUniquePtr>>(heaped_exceptions);
}

rust::String mol_sanitize_exception_type(const MolSanitizeExceptionUniquePtr &mol_except) {
	return mol_except->getType();
}

unsigned int atom_sanitize_exception_get_atom_idx(const MolSanitizeExceptionUniquePtr &mol_except) {
	MolSanitizeException *mol_except_ptr            = mol_except.get();
	AtomSanitizeException *atom_sanitize_except_ptr = (AtomSanitizeException *)mol_except_ptr;

	return atom_sanitize_except_ptr->getAtomIdx();
}

unsigned int get_num_atoms(const std::shared_ptr<ROMol> &mol, bool only_explicit) {
	return mol->getNumAtoms(only_explicit);
}

Atom &get_atom_with_idx(std::shared_ptr<ROMol> &mol, unsigned int idx) { return *mol->getAtomWithIdx(idx); }

rust::String get_symbol(const Atom &atom) { return atom.getSymbol(); }

bool get_is_aromatic(const Atom &atom) { return atom.getIsAromatic(); }
int get_atomic_num(const Atom &atom) { return atom.getAtomicNum(); }
int get_formal_charge(const Atom &atom) { return atom.getFormalCharge(); }
unsigned int get_total_num_hs(const Atom &atom) { return atom.getTotalNumHs(); }
unsigned int get_total_valence(const Atom &atom) { return atom.getTotalValence(); }
void set_formal_charge(Atom &atom, int what) { atom.setFormalCharge(what); }
void set_num_explicit_hs(Atom &atom, int what) { atom.setNumExplicitHs(what); }
void atom_update_property_cache(Atom &atom, bool strict) { atom.updatePropertyCache(strict); }

using HybridizationType = Atom::HybridizationType;
void atom_set_hybridization(Atom &atom, HybridizationType what) { atom.setHybridization(what); }
HybridizationType atom_get_hybridization(const Atom &atom) { return atom.getHybridization(); }

void ro_mol_update_property_cache(std::shared_ptr<ROMol> &mol, bool strict) { mol->updatePropertyCache(strict); }

// We add the different properties of and an atom to the Atom struct
void set_int_prop(Atom &atom, const std::string &key, int value) { atom.setProp(key, value); }
int get_int_prop(const Atom &atom, const std::string &key) { return atom.getProp<int>(key); }
void set_float_prop(Atom &atom, const std::string &key, double value) { atom.setProp(key, value); }
double get_float_prop(const Atom &atom, const std::string &key) { return atom.getProp<double>(key); }
void set_bool_prop(Atom &atom, const std::string &key, bool value) { atom.setProp(key, value); }
bool get_bool_prop(const Atom &atom, const std::string &key) { return atom.getProp<bool>(key); }
void set_prop(Atom &atom, const std::string &key, const std::string &value) { atom.setProp(key, value); }
rust::String get_prop(const Atom &atom, const std::string &key) { return atom.getProp<std::string>(key); }

unsigned int get_num_radical_electrons(const Atom &atom) { return atom.getNumRadicalElectrons(); }
} // namespace RDKit